scene/resources/texture.cpp

```diff
@@ -1867,6 +1868,237 @@ Ref<Image> GradientTexture::get_image() const {
 	return RenderingServer::get_singleton()->texture_2d_get(texture);
 }
 
+GradientTexture2D::GradientTexture2D() {
+	_queue_update();
+}
+
+GradientTexture2D::~GradientTexture2D() {
+	if (texture.is_valid()) {
+		RS::get_singleton()->free(texture);
+	}
+}
+
+void GradientTexture2D::set_gradient(Ref<Gradient> p_gradient) {
+	if (gradient == p_gradient) {
+		return;
+	}
+	if (gradient.is_valid()) {
+		gradient->disconnect(CoreStringNames::get_singleton()->changed, callable_mp(this, &GradientTexture2D::_queue_update));
+	}
+	gradient = p_gradient;
+	if (gradient.is_valid()) {
+		gradient->connect(CoreStringNames::get_singleton()->changed, callable_mp(this, &GradientTexture2D::_queue_update));
+	}
+	_queue_update();
+}
+
+Ref<Gradient> GradientTexture2D::get_gradient() const {
+	return gradient;
+}
+
+void GradientTexture2D::_queue_update() {
+	if (update_pending) {
+		return;
+	}
+	update_pending = true;
+	call_deferred("_update");
+}
+
+void GradientTexture2D::_update() {
+	update_pending = false;
+
+	if (gradient.is_null()) {
+		return;
+	}
+	Ref<Image> image;
+	image.instantiate();
+
+	if (gradient->get_points_count() <= 1) { // No need to interpolate.
+		image->create(width, height, false, Image::FORMAT_RGBA8);
+		Color color;
+		if (gradient->get_points_count() == 1) {
+			color = gradient->get_color(0);
+		} else {
+			color = Color(0, 0, 0, 1);
+		}
+		image->fill(color);
+	} else {
+		Vector<uint8_t> data;
+		data.resize(width * height * 4);
+		{
+			uint8_t *wd8 = data.ptrw();
+			Gradient &g = **gradient;
+			for (int y = 0; y < height; y++) {
+				for (int x = 0; x < width; x++) {
+					float ofs = _get_gradient_offset_at(x, y);
+					const Color &c = g.get_color_at_offset(ofs);
+
+					wd8[(x + (y * width)) * 4 + 0] = uint8_t(CLAMP(c.r * 255.0, 0, 255));
+					wd8[(x + (y * width)) * 4 + 1] = uint8_t(CLAMP(c.g * 255.0, 0, 255));
+					wd8[(x + (y * width)) * 4 + 2] = uint8_t(CLAMP(c.b * 255.0, 0, 255));
+					wd8[(x + (y * width)) * 4 + 3] = uint8_t(CLAMP(c.a * 255.0, 0, 255));
```

Comment:
Casting like `uint8_t(...)` will trim the fractional part which means that with the current implementation only values `>= 1.0` would result in `255` value (so for example `0.999` would result in `254`). I doubt that's desired. So probably it should be:
```suggestion
					wd8[(x + (y * width)) * 4 + 0] = uint8_t(CLAMP(c.r * 256.0, 0, 255));
					wd8[(x + (y * width)) * 4 + 1] = uint8_t(CLAMP(c.g * 256.0, 0, 255));
					wd8[(x + (y * width)) * 4 + 2] = uint8_t(CLAMP(c.b * 256.0, 0, 255));
					wd8[(x + (y * width)) * 4 + 3] = uint8_t(CLAMP(c.a * 256.0, 0, 255));
```
It would divide <0.0, 1.0> range into 256 equal length ranges (up to the precision errors) so for example values from range (more or less) <0.99609375, 1.0> would result in 255 value.

Reply:
Yeah but this is how `Image` does it currently.

Perhaps have to use `round()` in those cases?: https://github.com/godotengine/godot/issues/47022#issuecomment-799356453 (it would be less efficient I presume).

Reply:
I would keep it like that at least for now since it is consistent with Image and GradientTexture(1D) and discuss/tackle this problem in a future PR (for all occurrences).

Reply:
> Yeah but this is how `Image` does it currently.

Hmm, then maybe it's indeed a desired behavior? Not sure. But yeah, in this PR it can be made consistent with the rest of the code base then (and potential change could be made in a different PR).

> Perhaps have to use `round()` in those cases?: [#47022 (comment)](https://github.com/godotengine/godot/issues/47022#issuecomment-799356453) (it would be less efficient I presume).

After rethinking using `round()` would indeed be better then what I proposed here, it would result in a smaller average absolute "conversion error". And performance wise I guess it shouldn't be a big difference, it's probably just an additional `+ 0.5` instruction before the truncation (but it probably should be benchmarked).


Anyway, I agree with both of you, it's out of scope of this PR.